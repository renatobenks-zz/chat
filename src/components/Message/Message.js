// @flow

import * as React from 'react';
import { hot } from 'react-hot-loader';
import styled, { css } from 'styled-components';

import Avatar from '../Avatar/Avatar';

const Wrapper = styled.div`
  display: flex;
  width: 85%;
  justify-content: center;
  flex-wrap: nowrap;
  align-items: flex-start;
  margin: 0;

  ${props =>
    props.isFromMe &&
    css`
      flex-direction: row-reverse;
      align-self: flex-end;
      ${!props.hasAvatar &&
        css`
          margin-right: calc(${props.theme.avatar.size} + 1rem);
        `};
    `};

  ${props =>
    !props.isFromMe &&
    css`
      flex-direction: row;
      align-self: flex-start;
      ${!props.hasAvatar &&
        css`
          margin-left: calc(${props.theme.avatar.size} + 1rem);
        `};
    `};
`;

const DialogWrapper = styled.div`
  display: flex;
  width: 100%;
  flex-flow: column wrap;
  justify-content: center;
  margin: calc(${props => props.theme.avatar.size} / ${props => (props.hasAvatar ? 2 : 4)}) 0 0;

  ${props =>
    props.isFromMe
      ? css`
          align-items: flex-end;
          margin-right: ${props => (!props.hasAvatar ? 0 : '1rem')};
        `
      : css`
          align-items: flex-start;
          margin-left: ${props => (!props.hasAvatar ? 0 : '1rem')};
        `};
`;

const Dialog = styled.div`
  display: flex;
  flex: 1;
  padding: 1rem;
  align-items: center;
  justify-content: flex-start;

  ${props =>
    props.isFromMe &&
    css`
      background: ${props => props.theme.palette.primary};
      border-radius: 1rem 0 1rem 1rem;
    `};

  ${props =>
    !props.isFromMe &&
    css`
      background: ${props => props.theme.palette.gray};
      border-radius: 0 1rem 1rem 1rem;
    `};
`;

const Text = styled.p`
  margin: 0;
  color: ${props => props.theme.palette.white};
  font-size: 14px;
`;

const Time = styled.span`
  color: ${props => props.theme.palette.black};
  font-size: 10px;
  margin: 0.25rem 0 0;
`;

type Props = {
  children: React.Element<any>,
  isFromMe?: ?boolean,
  label?: string,
  avatar?: string,
};

const Message = ({
  children, isFromMe, label, avatar 
}: Props) => {
  const hasAvatar = !!avatar;

  return (
    <Wrapper isFromMe={isFromMe} hasAvatar={hasAvatar}>
      {avatar && <Avatar avatar={avatar} />}
      <DialogWrapper isFromMe={isFromMe} hasAvatar={hasAvatar}>
        <Dialog isFromMe={isFromMe}>
          <Text>{children}</Text>
        </Dialog>
        <Time>{label}</Time>
      </DialogWrapper>
    </Wrapper>
  );
};

export default hot(module)(Message);
