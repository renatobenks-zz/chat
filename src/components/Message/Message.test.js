import React from 'react';

import { render } from '../../../test/helpers';

import Message from './Message';

const renderer = component => render(component, { withTheme: true });

it('message with not children', () => {
  expect(renderer(<Message />)).toMatchSnapshot();
});

it('message with children', () => {
  expect(renderer(<Message>some message</Message>)).toMatchSnapshot();
});

it('message with label', () => {
  expect(renderer(<Message label="some label">some message</Message>)).toMatchSnapshot();
});

it('message from me', () => {
  expect(renderer(<Message isFromMe={true}>some message</Message>)).toMatchSnapshot();
});

it('message with avatar', () => {
  expect(renderer(<Message avatar={'image.png'}>some message</Message>)).toMatchSnapshot();
});

it('message with avatar from me', () => {
  expect(
    renderer(
      <Message isFromMe avatar={'image.png'}>
        some message
      </Message>,
    ),
  ).toMatchSnapshot();
});
