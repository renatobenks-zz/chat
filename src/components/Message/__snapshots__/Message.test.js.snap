// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message from me 1`] = `
<div
  className="sc-bdVaJa gXWJqC"
>
  <div
    className="sc-bwzfXH kmfMKZ"
  >
    <span
      className="sc-htpNat drENqL"
    >
      some message
    </span>
  </div>
  <span
    className="sc-bxivhb TwBys"
  />
</div>
`;

exports[`message with children 1`] = `
<div
  className="sc-bdVaJa ggSTVf"
>
  <div
    className="sc-bwzfXH dqrULd"
  >
    <span
      className="sc-htpNat drENqL"
    >
      some message
    </span>
  </div>
  <span
    className="sc-bxivhb TwBys"
  />
</div>
`;

exports[`message with label 1`] = `
<div
  className="sc-bdVaJa ggSTVf"
>
  <div
    className="sc-bwzfXH dqrULd"
  >
    <span
      className="sc-htpNat drENqL"
    >
      some message
    </span>
  </div>
  <span
    className="sc-bxivhb TwBys"
  >
    some label
  </span>
</div>
`;

exports[`message with not children 1`] = `
<div
  className="sc-bdVaJa ggSTVf"
>
  <div
    className="sc-bwzfXH dqrULd"
  >
    <span
      className="sc-htpNat drENqL"
    />
  </div>
  <span
    className="sc-bxivhb TwBys"
  />
</div>
`;
